### ISSUE TITLE
Consistent approach for propagating validation errors from GEOS

### ISSUE THREAD

**Issue by brendan-ward** [pygeos/pygeos#150]
This came out of discussion from #144 

In short, it would be good for pygeos to have a consistent approach to propagating validation errors from GEOS.

In some cases, GEOS returns fairly clear exception messages, e.g., passing an out of range number to `densify` parameter in `GEOSFrechetDistanceDensify_r`:

```
pygeos.GEOSException: IllegalArgumentException: Fraction is not in range (0.0 - 1.0]
```

In other cases, it might not.  My personal observation is that lack of validation of user inputs results in a not-insignificant maintenance burden for open source projects; bug reports uncover these gaps over time rather than anticipating them upfront.  That said, it is entirely possible to overdo validation, and we have lots of other things to focus our attention on here.

The latest guidance from @caspervdw is to avoid replicating validation and exceptions from GEOS if they are good enough, to reduce complexity in pygeos.

There are a couple of ways we could approach this:

**1. Use GEOS exceptions but ensure these are good enough within our tests.**
All this would require is that we add enough tests to our suite to check for invalid inputs, and ensure that the error messages are sufficiently clear - and remain consistent over GEOS versions.

If we discover that some invalid inputs result in segfaults or poor exception messages from GEOS, we should handle those on a case-by-case basis, including possibly adding a validation check in pygeos prior to calling GEOS code.

_pros:_
* simplicity
* avoid replicating validation logic and associated maintenance burdens

_cons:_
* inconsistent exception types: sometimes invalid inputs may result in `pygeos.GEOSException`, `ValueError`, or other exception types we define in pygeos for invalid inputs.
* outsources validation to GEOS, which means some risks of segfaults for unexpected inputs

**2. Add validation checks and raise appropriate exceptions in pygeos**
This would require anticipating a wide variety of invalid inputs and testing those specifically.

_pros:_
* consistent validation checks and associated exception types / messages
* ensures that `pygeos.GEOSException` errors are reserved for internal use or indicate exceptions that we should be catching first through validation in pygeos before calling GEOS.

_cons:_
* more effort to validate inputs and create associated tests
* replicates validation logic in GEOS

**Comment by jorisvandenbossche** [comment]
I am also very much in favor of letting GEOS handle it, if it does, and not replicate validation in `pygeos`. 

> If we discover that some invalid inputs result in segfaults or poor exception messages from GEOS, we should handle those on a case-by-case basis, including possibly adding a validation check in pygeos prior to calling GEOS code.

It might certainly be needed to have additional checks in PyGEOS for such cases (because in theory our users should never experience a segfault). But in general we should *also* report those to GEOS, so they can potentially get fixed there.


---

We should maybe make the distinction between validating geometries and validating non-geometry keywords. For invalid geometries, I think seeing GEOSExceptions for those is fine, and in general I don't think we should start validating geometries. 
But for other keywords, we also already do conversion from python types to the C types GEOS expects, and thus can easily do some validation there. For those I would be less opposed to have some more validation checks in Python.

**Comment by caspervdw** [comment]
Thanks for writing up this issue @brendan-ward . I prefer leaving the error validation to GEOS, unless in the following cases (your first option). This at the expense of dropping the python-specific exception types. Cases that do need validation from pygeos:

 - segfaults: they must be avoided by validating on the C level
 - handling of missing values: This is a distinct feature of pygeos that is not supported by GEOS. Missing values (`None` for geometries, `NaN` for floats) need custom handling on the C level.

There are some specific points raised in #113 that I want to discuss. In the following cases I transformed a "wrong" input into a `None` result, while I could also have set an error:

 - GetPointN, GetInteriorRingN, GetGeometryN when N is out of bounds or the geometry is of incorrect type
 - GetExteriorRing if geometry is not a polygon
 - GEOSBoundaryAllTypes_r when the geometry is a collection

Which of the two is better? Return `None` or set an exception?

**Comment by chaitan94** [comment]
> Which of the two is better? Return None or set an exception?

I would say returning `None` is a good idea for a library providing vectorized functions. However, in some cases, an exception might be more preferable. And I think we can't decide on one single approach. So maybe one thing we could do is expose this as an option to the user, but defaulting to returning `None`.

So, for example, `get_point` function could be defined with an additional parameter called `fail_on_wrong_inputs`. When this is True, we throw instead of returning `None`:

```python
def get_point(geometry, index, fail_on_wrong_inputs=False):
```

Depending on how detailed control we want to expose this could be a bit more structured. For example, using, say, bit flags.

```python
# Types of wrong inputs which could cause failure
FAIL_NEVER                   = 0b0000
FAIL_ON_UNSUPPORTED_GEOMETRY = 0b0001
FAIL_ON_OUT_OF_BOUNDS        = 0b0010
FAIL_ALWAYS = FAIL_ON_UNSUPPORTED_GEOMETRY|FAIL_ON_OUT_OF_BOUNDS

def get_point(geometry, index, fail_on=FAIL_NEVER):
    # imagine implementation here

# Default behaviour - doesn't fail, passes on `None` instead
get_point(line_with_3_points, [0, 1, 9])

# Fail only on out of bounds
get_point(line_with_3_points, [0, 1, 9], fail_on=FAIL_ON_OUT_OF_BOUNDS)

# Fail in both cases
get_point(line_with_3_points, [0, 1, 9], fail_on=FAIL_ALWAYS)
```

Looks a bit verbose though. But do let me know what you guys think.

**Comment by brendan-ward** [comment]
In terms of prior art, I found some examples to consider here:
* [`pandas.drop()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop.html) has a parameter `errors` which is either `"ignore"` or `"raise"` defaulting to `"raise"`.  Something like this seems like a reasonable approach where there is ambiguity about when we should raise errors; this prevents them from occurring silently unless explicitly silenced by user.
* various numpy functions (e.g., `axis` in [`fft`](https://numpy.org/doc/1.18/reference/generated/numpy.fft.fft.html#numpy.fft.fft)) raise exceptions if one of the non-vector parameters (or constrained vector in the case of [`digitize`](https://numpy.org/doc/1.18/reference/generated/numpy.digitize.html#numpy.digitize)) is not valid for the inputs.  This seems similar to what we may want to do for `GetPointN`, etc - though I appreciate that these parameters may be scalars or vectors. 
* `mode` in [`numpy.take()`](https://numpy.org/doc/1.18/reference/generated/numpy.take.html#numpy.take) allows you to raise errors or have default behavior, e.g., when indexes are out of bounds, clip them to the valid range.

I'm not finding good examples where invalid inputs to numpy / scipy will return equivalent of `None` when the parameters are invalid, but it would be good to compare these to the above if there are any.

**Comment by jorisvandenbossche** [comment]
I don't know if the current behaviour was already inspired on it, but so what we current have in geopandas:

```
import shapely.geometry
import geopandas
s = geopandas.GeoSeries([shapely.geometry.Point(1, 1), shapely.geometry.box(1, 1, 3, 3)])

In [30]: s.exterior
Out[30]: 
0                                                 None
1    LINEARRING (3.00000 1.00000, 3.00000 3.00000, ...
dtype: geometry
```

which allows to call `exterior` even if you have mixed geometry types (not being able to do this might get annoying if you ended up with some points or lines in your polygon dataset, eg through some spatial operation). 

We could also check what PostGIS does?

**Comment by jorisvandenbossche** [comment]
Yes, so Postgis also returns NULL for functions like this that require a specific geometry type, see eg https://postgis.net/docs/ST_ExteriorRing.html

**Comment by caspervdw** [comment]
I think this discussion came to a consensus. Maybe we could start keeping these design decisions somewhere? @jorisvandenbossche  @brendan-ward  do you have any experience with that?

**Comment by jorisvandenbossche** [comment]
I think a "developers" / "internals" / "design" document to capture some of those discussions would indeed be useful (eg also in general describing the missing values handling, the GEOS context handling, ..). Putting such a page in the docs would be fine I think
